plugins/circleci/src/hooks/usePipelines.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2020 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { Pipeline, PipelineInfo, Workflow } from '../types';
+import { errorApiRef, useApi } from '@backstage/core-plugin-api';
+import { AuthenticationError } from '@backstage/errors';
+import useAsyncRetry from 'react-use/lib/useAsyncRetry';
+import { useProjectSlugFromEntity } from './useProjectSlugFromEntity';
+import { circleCIApiRef } from '../api';
+
+type PipelinesPage = {
+  items?: PipelineInfo[];
+  pageToken?: string;
+  nextPageToken?: string;
+};
+
+export function usePipelines() {
+  const { repo, owner, projectSlug } = useProjectSlugFromEntity();
+  const api = useApi(circleCIApiRef);
+  const errorApi = useApi(errorApiRef);
+  const pipelinesPage = useRef<PipelinesPage>({ items: [] });
+  const [pageToken, setPageToken] = useState<string>();
+  const [pipelines, setPipelines] = useState<PipelineInfo[]>([]);
+
+  const getWorkflows = useCallback(
+    async (pipelineId: string) => {
+      try {
+        const response = await api.getWorkflowsForPipeline(pipelineId);
+        return response.items;
+      } catch (e) {
+        if (!(e instanceof AuthenticationError)) {
+          errorApi.post(e);
+        }
+        return Promise.reject(e);
+      }
+    },
+    [api, errorApi],
+  );
+
+  const transform = async (
```

Comment:
If i'm reading things right, there's a lot of unbounded mapping and transformation that contains API calls in this file. So you'd in theory hammer circleci with an unlimited number of concurrent calls which they may not approve of. We often try to wrap stuff like this in a p-limit that bounds the concurrency to a small number. See [here](https://github.com/backstage/backstage/blob/55725922a5d149ed6a5bb0d5976ec3130b14dd96/plugins/catalog-backend-module-msgraph/src/microsoftGraph/read.ts#L94) for an example

Reply:
Using `p-limit` to limit concurrency.